rfcs/0048-BYOC-Marvell-ML-accelerator-integration.md

```diff
@@ -0,0 +1,547 @@
+      %2 = @tvmgen_mrvl_main_2(%1, /* en_id=4138 */) /* ty=Tensor[(1, 14, 14, 64), float32] */;
+      %3 = @tvmgen_mrvl_main_3(%2, /* en_id=4139 */) /* ty=Tensor[(1, 14, 14, 32), float32] */;
+      %4 = @tvmgen_mrvl_main_4(%3, /* en_id=4140 */) /* ty=Tensor[(1, 7, 7, 32), float32] */;
+
+      def @tvmgen_mrvl_main_0(%mrvl_0_i0: Tensor[(1, 1, 28, 28), float32], Inline=1, Compiler="mrvl",
+          global_symbol="tvmgen_mrvl_main_0", Primitive=1) -> Tensor[(1, 28, 28, 1), float32] {
+        layout_transform(%mrvl_0_i0, src_layout="NCHW", dst_layout="NHWC",
+            /* en_id=3334 */) /* ty=Tensor[(1, 28, 28, 1), float32] */
+      }
+```
+
+* Currently, in order for the following Marvell classes/functions to identify a Mrvl subgraphs and a non-Mrvl
+  subgraph from the layout-converted, composited/fused IR graph, we are utilizing the unique en_id attribute
+  stored for the Class CallNode and the class Tuple (include/tvm/relay/expr.h).
+    * in mrvl.py: class MrvlIRGraphUtils.RestOfMrvlLayers(ExprMutator) is used to convert the non-Mrvl subgraph,
+      which can have composited Marvell layer(s) back to their original IR nodes (e.g., to use original tensor
+      layout and with no compositions)
+    * in mrvl.py: class MrvlIRGraphUtils.RestMrvlLayersGetInputs(ExprVisitor) is used to reconstruct the input
+      tensor for the non-Mrvl subgraph so that it become a IR graph, which is recognized by the TVM LLVM build.
+    * in mrvl.py: the revert_mrvl_mod_to_orig() function is defined to convert the initial non-Mrvl subgraph back
+      to a IR subgraph using original layouts with no Marvell-specific compositions (e.g., similar to what was
+      given by the frontend)
+
+```
+def revert_mrvl_mod_to_orig(mod_mrvl_subgraph, mrvl_layers_in_mrvl_subgraph, debug=False):
+    """
+
+    def run_opt_pass(mod, passes):
+        passes = passes if isinstance(passes, list) else [passes]
+        seq = tvm.transform.Sequential(passes)
+        with tvm.transform.PassContext(opt_level=3):
+            mod = seq(mod)
+        return mod
+
+    mod_new = tvm.IRModule(mod_mrvl.functions, mod_mrvl.type_definitions)
+    mod_new["main"] = MrvlSubgraphToRevert(mrvl_layers_in_mrvl_subgraph, mod_mrvl).visit(mod_mrvl["main"])
+    mod_new = relay.transform.RemoveUnusedFunctions()(mod_new)
+    mod_new = relay.transform.InferType()(mod_new)
+    mod_new = run_opt_pass(mod_new, relay.transform.DefuseOps())
+    mod_new = run_opt_pass(mod_new, relay.transform.ConvertLayout({"nn.conv2d": ["NCHW", "OIHW"], "nn.max_pool2d": ["NCHW"]}))
+    mod_new = run_opt_pass(mod_new, relay.transform.SimplifyExpr())
+    mod_new = run_opt_pass(mod_new, relay.transform._ffi_api.DropNoopTranspose())
+    mod_new = run_opt_pass(mod_new, relay.transform.InferType())
+    return mod_new
+```
+
+* Marvell-specific graph executor codegen, We have defined call backs and extension functions in the following files:
```

Comment:
could you motivate this further? it's hard to understand why you need to output your own JSON format without some explanation here.

Reply:
oh the email version of this question was linking to a different RFC segment (the light green section above) -- led to me to answer differently in my in-line reply to your email.
Sorry and since I noticed here and can see the correct light green block above corresponding to your question, so let me reply here again properly.

We are using the TVM graph executor codegen to process BYOC-Marvell-relay-seq-generated Marvell-part of the IR sub-graph, which includes Marvell-specific GraphInputNode object(s) & attributes and Marvell-specific GraphOpNode object(s) and attributes. When processing the Marvell sub-graph in the TVM graph executor codegen, we need to specialize the generation code in order to dump extra Marvell-specified attributes to node-JSON file (also in a more readable format). The original code can't do what we need; hence, we are using derive classes and call back functions in C++ to overwrite defaults here.